arangod/RocksDBEngine/RocksDBPersistedLog.cpp

```diff
@@ -0,0 +1,290 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+#include <Basics/Exceptions.h>
+#include <Basics/RocksDBUtils.h>
+#include <Basics/ScopeGuard.h>
+#include <Basics/debugging.h>
+
+#include "RocksDBKey.h"
+#include "RocksDBPersistedLog.h"
+#include "RocksDBValue.h"
+
+using namespace arangodb;
+using namespace arangodb::replication2;
+
+RocksDBPersistedLog::RocksDBPersistedLog(replication2::LogId id, uint64_t objectId,
+                                         std::shared_ptr<RocksDBLogPersistor> persistor)
+    : PersistedLog(id), _objectId(objectId), _persistor(std::move(persistor)) {}
+
+auto RocksDBPersistedLog::insert(LogIterator& iter, WriteOptions const& options) -> Result {
+  rocksdb::WriteBatch wb;
+  auto res = prepareWriteBatch(iter, wb);
+  if (!res.ok()) {
+    return res;
+  }
+
+  if (auto s = _persistor->_db->Write({}, &wb); !s.ok()) {
+    return rocksutils::convertStatus(s);
+  }
+
+  if (options.waitForSync) {
+    // At this point we have to make sure that every previous log entry is
+    // synced as well. Otherwise we might get holes in the log.
+    if (auto s = _persistor->_db->SyncWAL(); !s.ok()) {
```

Comment:
I am not sure if `SyncWAL()` is portable. At least it seemed that on Windows it always returned an error message such as "unsupported". Please refer to https://github.com/arangodb/arangodb/blob/devel/arangod/RocksDBEngine/RocksDBSyncThread.cpp#L78-L87

Reply:
This would then imply that we do not support `waitForSync` on windows?

Reply:
At least not by calling `SyncWAL()`. I think on Windows the supported way would be to set `sync` in the `WriteOptions`.

Reply:
Turns out Windows supports `SyncWal` now.